### ISSUE TITLE
functools wraps is not handled correctly w.r.t return value type

### ISSUE THREAD

**Issue by jonathanunderwood** [microsoft/pyright#10436]
**Describe the bug**

Consider the following simple example where I am trying to wrap a pre-existing decorator from a third party package while maintaining the function signature.

```python
from functools import wraps
import dagster as dg

@wraps(dg.asset)
def asset(*args, **kwargs): 
    # do something with kwargs here - removed for brevity
    return dg.asset(*args, **kwargs)

@asset
def daily_sales() -> None: ...

defs = dg.Definitions(assets=[daily_sales])
```

Running pyright on this file:

```bash
pyright pyright_test.py
pyright_test.py
  pyright_test.py:16:31 - error: Argument of type "list[() -> None]" cannot be assigned to parameter "assets" of type "Iterable[AssetsDefinition | AssetSpec | SourceAsset | CacheableAssetsDefinition] | None" in function "__new__"
    Type "() -> None" is not assignable to type "AssetsDefinition | AssetSpec | SourceAsset | CacheableAssetsDefinition"
      "function" is not assignable to "AssetsDefinition"
      "function" is not assignable to "AssetSpec"
      "function" is not assignable to "SourceAsset"
      "function" is not assignable to "CacheableAssetsDefinition" (reportArgumentType)
1 error, 0 warnings, 0 informations
```

My expectation here is that the return type of `asset()` is set by `functools.wraps` to be that of the wrapped `dagster.asset` function, i.e. `AssetsDefinition`, and that pyright would reflect that and not raise an error.

If I explicitly set the return type of `asset()`, then pyright does not raise an error. I.e.

```python
from functools import wraps
import dagster as dg

@wraps(dg.asset)
def asset(*args, **kwargs) -> dg.AssetsDefinition:
    # do something with kwargs here
    return dg.asset(*args, **kwargs)

@asset
def daily_sales() -> None: ...

defs = dg.Definitions(assets=[daily_sales])
```

```bash
pyright pyright_test.py
0 errors, 0 warnings, 0 informations
```

**Code or Screenshots**
See above

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

Command line tool. 

```bash
pyright --version
pyright 1.1.400
```

**Comment by jonathanunderwood** [comment]
I've reduced this down to a reproducer that doesn't require the dagster package below:

```python
from functools import wraps
from typing import Any, Callable, overload, Union

class AssetsDefinition:
    def __init__(self, fn): ...


@overload
def dg_asset(
    fn: Callable[..., Any],
    **kwargs,
) -> AssetsDefinition: ...

@overload
def dg_asset(*, name="blah", **kwargs) -> Callable[[Callable[..., Any]], AssetsDefinition]:
    ...


def dg_asset(fn: Callable[..., Any] | None = None, name: str| None = None, **kwargs) -> Union[AssetsDefinition, Callable[[Callable[..., Any]], AssetsDefinition]]:
    return AssetsDefinition(fn)

def definitions(assets: list[AssetsDefinition]) -> list[Any]:
    ...

@wraps(dg_asset)
def asset(*args, **kwargs): # -> AssetsDefinition:
    # do something with kwargs here
    return dg_asset(*args, **kwargs)

@asset
def daily_sales() -> None: ...

defs = definitions(assets=[daily_sales])
```
[pyright playground link](https://pyright-play.net/?code=GYJw9gtgBMCuB2BjALmMAbAzlAlhADmCMlAO4gCG%2BmAUKJFMgJ747wDmuBRJAgvEwA0UAMIV06CgCN0AU2FgAbrJDowFACbCAqvBxh4NGokmZsvM7OSYAIrOBscyffABcNKJ6gb7UAPp%2BjsgBABSYsujAwsDwAJSuUAB0yUYmFGZQABruXkkpNAACSipqmjQ%2BwN7sfunhyCEeXjEJYhLScgDayYnC-EwAuoKNngBUIwDWpBQg7JhDsVAAtAB8UBZ1tvZBLgndRkXKquoa5b4a1bVWISPC8BQQsgC8AEQyFAAWz8Jjk9OzCytROJJDJZB0Oq0QZ1ur0BP1BmtLNY7A49M4DP0cl49kYKlUakiQs0gW1QV1krCBlAAD5QAByBlkUEe9MZt3usgSmGQIFpDPgTJZ-PkUB%2BUxmmABq10Lg66ysm1RThcwkh7TBEOB6vJPTWcIR8uRWzRLnhWM8ICssBA8ERGxR2wMRLiuLOxuVBkwIUu1gS6Bw3LlSMVjvg-SlUH9gb6mOGeUS%2B3IVC95wJdVip0qPuufzmoom4v%2BCQAxEtVoaQyaDOaoKWNGAoJhIFZ3mxOKQnO8oL8JVB3ipZHHLchrbbU9mRrnvgXcxnCj7M94KDh0Ew-JhxLIvRHhbt8hVsCyKqGvT7MI8Ohpl6v15vMOGgA)


Interestingly, to reproduce the issue, did require the use of overload.

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

In your second example, you haven't provided a return type annotation for `asset`, so pyright attempts to infer it based on the return statement. You also haven't provided annotations for the `args` or `kwargs` parameters, so they are assumed to be `Unknown`. The call expression `dg_asset(*args, **kwargs)` targets an overload, so the rules for overload call evaluation apply. You can find those documented [here](https://typing.python.org/en/latest/spec/overload.html#overload-call-evaluation) in the typing spec. Applying these rules, the call evaluates to `Unknown`. That means the final signature of `asset` (prior to applying the `wraps` decorator`) is `(...) -> Unknown`.

When pyright encounters a function decorator whose type evaluates to `(...) -> Unknown`, it assumes that the decorator does not affect the signature of the decorated function. This is a good assumption under most circumstances, and it typically works well for untyped code bases, since it preserves completion suggestion behavior. In your case, this behavior work against you because you actually want the signature to be changed based on the decorator.

You can fix this in your code by adding some explicit type annotations. For example, you can uncomment the `-> AssetsDefinition` annotation for the `asset` function.

**Comment by jonathanunderwood** [comment]
Thanks for the quick response @erictraut, and the detailed explanation - I very much appreciate that you took the time to respond even though this is not a bug in the project. The part I was missing in all of this was how type checking/inference works for overloads which you linked to.

For anyone reding this in the future, another informative thread is: https://discuss.python.org/t/taking-the-argument-signature-from-a-different-function/42618/6
